File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.cpp
class Solution {
public:
    void findNSE(vector<int>& nse, int n, vector<int> arr) {
        stack<int> st;
        for(int i=n-1; i>=0; i--) {
            while(!st.empty() && arr[st.top()] >= arr[i]) {
                st.pop();
            }
            nse[i] = st.empty() ? n : st.top();
            st.push(i);
        }
    }
    void findPSE(vector<int>& pse, int n, vector<int> arr) {
        stack<int> st;
        for(int i=0; i<n; i++) {
            while(!st.empty() && arr[st.top()] > arr[i]) {
                st.pop();
            }
            pse[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }
    }

    int sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        vector<int> nse(n);
        vector<int> pse(n);
        findNSE(nse, n, arr);
        findPSE(pse, n, arr);
        long long sum = 0;
        const int mod = 1e9 + 7;

        for (int i = 0; i < n; i++) {
            long long left = nse[i]-i;
            long long right = i-pse[i];
            sum = (sum + left * right % mod * arr[i]) % mod;
        }

        return (int)sum;
    }
};

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    int[] findNse(int[] arr) {
        Stack<Integer> st = new Stack<>();
        int[] ans = new int[arr.length];
        for(int i=arr.length-1; i>=0; i--) {
            int curr = arr[i];
            while(!st.empty() && arr[st.peek()] >= curr) st.pop();
            if(st.empty()) {
                ans[i] = arr.length;
            } else {
                ans[i] = st.peek();
            }
            st.push(i);
        }
        return ans;
    }
    int[] findPse(int[] arr) {
        Stack<Integer> st = new Stack<>();
        int[] ans = new int[arr.length];
        for(int i=0; i<arr.length; i++) {
            int curr = arr[i];
            while(!st.empty() && arr[st.peek()] > curr) st.pop();
            if(st.empty()) {
                ans[i] = -1;
            } else {
                ans[i] = st.peek();
            }
            st.push(i);
        }
        return ans;
    }
    public int sumSubarrayMins(int[] arr) {
        int nse[] = findNse(arr);
        int pse[] = findPse(arr);
        int sum = 0;
        int mod = (int)1e9 + 7;
        for(int i=0; i<arr.length; i++) {
            int left = i-pse[i];
            int right = nse[i]-i;
            long cont = left*right*1L;
            int val = (int)((cont * arr[i]) % mod);
            sum = (sum+val)%mod;
        }
        return sum;
    }
}